### DISCUSSION TITLE
Integrate JWT token for credential response

### DISCUSSION THREAD

**Discussion by philippedasilva-orizone** [wavezync/vue3-google-signin#27] [Ideas]
Hey,

I thought we could all benefit from getting a decrypted jwt token in the CredentialResponse type instead of the actual string returned by Google on successfully logging the user.

I know it would probably add a dependency but I believe we, end users of the component, would anyway still have to add it.

My own implementation uses the universal-base64Url package (https://www.npmjs.com/package/universal-base64url):
```typescript
      const base64Url = jwt.split('.')[1];
      const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
      const jsonPayload = decodeURIComponent(
        window
          .atob(base64)
          .split('')
          .map(function (c) {
            return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
          })
          .join('')
      );
      const decodedToken = JSON.parse(jsonPayload);
      console.log(decodedToken);
```

Ideally, I would actually get the CredentialResponse with a typed "credential" so I can directly benefit from my IDE intellisense with a `decodedToken.email` or `decodedToken.family_name`...

Any reason not to do it?

**Comment by kasvith** [comment]
We thought to keep this as possibly closer to the underlying google library. But your point is quite valid. Lot of people ask how to use it.

I thought of two ways

- Adding this to another library(vite now supposed to support multiple library entry points, so we can make it something like `import {decodeToken} from 'vue3-google-signin/utils'`)
- Adding a FAQ and using your example there as a guide

**Reply by philippedasilva-orizone** [reply]
The FAQ approach is probably the minimum that should be done.

I admit that I have some doubts on adding it to another library for the following reasons:
a/ It'll probably add another dependency as it makes no sense to implement base64url enc/dec
b/ It adds a functional dependency on the google identity service as if Google decides to change the token structure, it would require to update the package

However, if it's not in your package, developers using it would still have to update their own code to decode the jwt token anyway and since they'll be using your wrapper on top of google identity service, they may have some difficulties to see where the error lies :p

It's your call and I'm happy to PR the approach the way you think it the best ;)

Another approach may be to actually implement this inside another util package such as VueUse? I could see this JWT token being retrieved using other packages than yours :p

**Reply by kasvith** [reply]
FAQ approach is what we can do in a short time.

But I ve in my mind to add this capability using a separate library entry point with vite as i mentioned. That way it will be added to the bundle only if you used it(with tree shaking).

Im bit waiting until vite is stabilizing multiple library entry points

**Reply by kasvith** [reply]
Vite upstream PR https://github.com/vitejs/vite/pull/7047

**Reply by kasvith** [reply]
@philippedasilva-orizone can you create an issue for this? we can track progress there. Once Vite stabilize API we will expose it.

I actually wanted to export some methods provided by the plugin to its own namespaces like `vue3-google-signin/oauth2` etc first place. Hope your method can be part of the ``vue3-google-signin/util`

**Reply by philippedasilva-orizone** [reply]
Shouldn't we just convert that discussion topic into an issue / feature request ? Not that I don't want to but during the week, I have less time to spend on my hobby projects :p

**Reply by kasvith** [reply]
Converted to issue #28

**Reply by kasvith** [reply]
Lets continue discussion there
